### PR TITLE
support escalation slack notification

### PR DESCRIPTION
**Pull request by fzhao99**
# BACKEND PULL REQUEST

## Related Issue

- Mostly resolves #[5723](https://github.com/CDCgov/prime-simplereport/issues/5723)
- To be followed up with a PR that adds the link to the page

## Changes Proposed

- Uses the Slack Webhooks API to send messages into Slack when hit
- Messages will be sent to the [#proj-sr-support-notifications channel](https://join.slack.com/share/enQtNTc0NDMyOTI1NDQ2Ny0zYTVkOGQyMzlkMTVlNjA1N2I4ODc1ZTRlMzE3NWVhODczZmIwMjU3NzgwZWQ3MmY5YjM3NmFiNThiNjY0MDlh)
- Final copy and design to be oked before merging

## Testing

- Deployed on dev5 under the new [escalate to engineering page](https://dev5.simplereport.gov/app/admin/escalate-to-engineering)

https://github.com/CDCgov/prime-simplereport/assets/29645040/582da4f9-e4e8-4cf9-b7c8-75df3cecaa5f

### CODE REVIEW

**Comment by jayna-SkylightDigital** [comment]
@fzhao99 This looks good to me -- I'd like to make sure we get the Support Team folks to understand how it works/how to access it/when to use it in our Customer Support Bi-Weekly Meeting on August 17th.

**Review by BobanL — COMMENTED** [review]
Left a couple of comments for some cleanup, but otherwise works great! Thanks for working on this 😄

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfig.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    try {
+      String postUrl = url + token;
+      String response = restTemplate.postForObject(postUrl, entity, String.class);
+
+      // Slack will respond with "ok" if the command was successful.
+      if (response.contains("ok")) return true;
+      log.error("Slack returned unexpected: " + response);
+      throw new SlackEscalationException();
+    } catch (RestClientException | SlackEscalationException e) {
+      return false;
```
[the remark this answers is not in the document]
Comment (fzhao99) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfig.java:48]: I'm a bit confused as to how I'd have `response` in scope here since the result of `postForObject` wouldn't be defined in the catch block. Am I missing something?

[the remark this answers is not in the document]
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfig.java]: Oops, you are right you wouldn't have that there. you could just include the RestClientException though
```java
    try {
      String postUrl = url + token; 
      restTemplate.postForObject(postUrl, entity, String.class);
      return true
    } catch (RestClientException e) {
      log.error("Slack returned unexpected: " + e);
      return false;
    }
```

**Review by BobanL — DISMISSED** [review]
Changes look good! Thank you!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackWebhookService.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class SlackWebhookService {
+
+  private SlackConfigService slackConfigService;
+
+  public boolean sendSlackEscalationMessage() {
```
Comment (johanna-skylight) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackWebhookService.java:12]: Should this method have the `@AuthorizationConfiguration.RequireGlobalAdminUser` annotation?

File: frontend/src/app/supportAdmin/Escalations/Escalations.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+
+import { useSendSupportEscalationMutation } from "../../../generated/graphql";
+import Button from "../../commonComponents/Button/Button";
+import { showSuccess } from "../../utils/srToast";
+import { useDocumentTitle } from "../../utils/hooks";
+
+export const Escalations = () => {
+  useDocumentTitle("Escalations");
```
Comment (johanna-skylight) [inline frontend/src/app/supportAdmin/Escalations/Escalations.tsx:9]: Should this use the `escalationPageTitle` in the pageTitles.tsx ?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  public boolean makeEscalationRequest() {
+    log.info("making escalation request");
+    String response = "";
+    try {
+      String postUrl = url + token;
+      response = restTemplate.postForObject(postUrl, entity, String.class);
+      if (response.contains("ok")) return true;
+      throw new RestClientException("Slack escalation didn't return ok response");
```
Comment (fzhao99) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java:43]: turns out `postForObject` won't throw an exception here in the event that we have an unsuccessful request: it'll just return null here and we'll have false positive success toasts. We need to explicitly check for the "ok" response to ensure we had a successful response.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfig.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzhao99) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfig.java]: turns out `postForObject` won't throw an exception here in the event that we have an unsuccessful request: it'll just return null and we'll have false positive success toasts if we just return true like we had it. I've added an explicit  check for the "ok" response and then throw a `RestClientException` manually if it doesn't. Let me know if this makes sense / there are other suggestions for improvement!

File: ops/dev5/_data.tf [hunk 5]
```diff
@@ -235,3 +235,8 @@ data "azurerm_key_vault_secret" "datahub_signing_key" {
   name         = "datahub-signing-key-test"
   key_vault_id = data.azurerm_key_vault.sr_global.id
 }
+
+data "azurerm_key_vault_secret" "slack_hook_token" {
```
Comment (emyl3) [inline ops/dev5/_data.tf:239]: should we also update this for the other lowers as well or will you be doing that in a separate PR?

File: ops/dev5/api.tf [hunk 6]
```diff
@@ -60,6 +60,8 @@ module "simple_report_api" {
     DATAHUB_API_KEY                               = "@Microsoft.KeyVault(SecretUri=${data.azurerm_key_vault_secret.datahub_api_key.id})"
     DATAHUB_FHIR_KEY                              = "@Microsoft.KeyVault(SecretUri=${data.azurerm_key_vault_secret.datahub_fhir_key.id})"
     DATAHUB_SIGNING_KEY                           = "@Microsoft.KeyVault(SecretUri=${data.azurerm_key_vault_secret.datahub_signing_key.id})"
+    SLACK_HOOK_TOKEN                              = "@Microsoft.KeyVault(SecretUri=${data.azurerm_key_vault_secret.slack_hook_token.id})"
```
Comment (emyl3) [inline ops/dev5/api.tf:63]: same question as above 🤔

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+@Slf4j
+public class SlackConfigService {
+  private final String escalationMessage = "A new support escalation has been received!";
+
+  private String url = "https://hooks.slack.com/services/";
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java:18]: since we are hardcoding this here -- should I go into the vault and clean up the `slack-hook-url-dev` that I added for you? let me know if this needs to be done! 😸

File: frontend/src/app/supportAdmin/Escalations/Escalations.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+          <form className="prime-container card-container">
+            <div className="usa-card__header">
+              <h1 className="font-heading-lg margin-top-0 margin-bottom-0">
+                Escalate to engineering
+              </h1>
+            </div>
+            <div className="usa-card__body margin-top-1">
+              <div className="grid-row grid-gap margin-top-2">
+                <div className="tablet:grid-col">
+                  <p>
```
Comment (emyl3) [inline frontend/src/app/supportAdmin/Escalations/Escalations.tsx:26]: nit: can we add the class `usa-prose` to this so we get some line height?

File: frontend/src/app/supportAdmin/Escalations/Escalations.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                    Use the below button to notify the SimpleReport engineering
+                    team about a tier 2 escalation. Make sure you've filled out
+                    information about the escalation in{" "}
+                    <a href="https://app.smartsheetgov.com/sheets/wwchRWw4jr55WpX8XwC2h9Mr9rVr3Q3hww64pW61">
```
Comment (emyl3) [inline frontend/src/app/supportAdmin/Escalations/Escalations.tsx:30]: nit: can we `target="_blank" rel="noopener noreferrer"` to this link?

**Review by emyl3 — COMMENTED** [review]
Left a couple comments! Awesome work on this, Bob!! Let me know what you decide to do 😸

File: backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java (hunk 7 above)
Comment (fzhao99) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/supportescalation/SlackConfigService.java:18]: I got access to my SU and deleted them myself!

File: ops/dev5/_data.tf (hunk 5 above)
Comment (fzhao99) [reply ops/dev5/_data.tf:239]: yeah maybe something worth getting other folks to chime in on, but I didn't want to add anything to the lower envs since there really shouldn't be a reason (other verifying things work as in this PR) for those secrets to be there. Wouldn't want a situation where we get a false positive from testing there. Plus, folks can always plug in the token locally if they need to configure something.

If folks agree, I'll probably go back and delete this token on dev5 after this goes in

Comment (emyl3) [reply ops/dev5/_data.tf:239]: This reasoning makes sense to me! Thank you, Bob!

**Review by emyl3 — APPROVED** [review]
LGTM!

**Review by BobanL — APPROVED** [review]

**Review by johanna-skylight — APPROVED** [review]
🚀
